elytron/WFCORE-4360-CredentialStore_Expression_Resolution.adoc

```diff
@@ -0,0 +1,420 @@
+ * `create` - If the credential store file does not exist should it be created?
+ * `populate` - If the credential store does not already contain an entry for the `default-alias` should one be added?
+ * `key-size` - The key size to use for any automatically generated `SecretKey`.
+ * `default-alias` - The name to use for any alias dynamically added to the credential store.
+
+The new resource will expose the `org.wildfly.security.credential-store` so can be referenced from other parts of the management model.  It should be noted use
+across the model would be limited as this store does not support password credentials.
+
+The configuration attributes for this credential store mean that the resource can be added to the configuration and as the server boots if the credential store
+does not exist it will be dynamically created and it can also be dynamically populated with a `SecretKey` - this means the resource is suitable for a very 
+simple out of the box getting started experience.
+
+==== Key manipulation - WildFly Elytron Tool
+
+It will be possible to generate, export, and import the SecretKey instances using the command line tool.  The same actions for secret key manipulation will be
+applicable to both the existing `KeyStoreCredentialStore` and the new `PreopertiesCredentialStore`, the only difference being the `PasswordCredentialStore`
+does not require a password to access it.
+
+The Elytron tool `credential-store` command already has an `--add` action, rather than overloading this action for different types of key and different
+operations which will become complex very quickly a new set of actions will be added:
+
+ * `--generate-secret-key`
+ * `--export-secret-key`
+ * `--import-secret-key`  
+
+The `--generate-secret-key` action will take the alias to store the key as an argument in the same pattern as `add` it will also take the following additional
+option:
+
+ * `--size` - The key size (number of bits) - acceptable values 128, 192, or 256 - default 256.
+ 
+The `--export-secret-key` action will take the alias of the previously stored key as an argument, no further option will be required.
+
+The `--export-secret-key` action will return a single Base64 value which is the key in it's encoded form.
+
+The `--import-secret-key` action will take the alias to store the key as an argument, it will also take the following additional option: -
+
+ * `--key` - The key to import encoded using Base64.
+
+If the `--key` option is not specified the user will be prompted to enter the key interactively, this will avoid the key being seen in the current running processes and avoid the key being cached in the users command line history.
+
+NOTE: For `generate-secret-key` and `import-secret-key` if an entry already exists for that alias / algorithm combination the behaviour will be the same as the `add` action in relation to replacing the entry or reporting an error.  (To be checked)
```

Comment:
I have tried the `add` action and it replaces an existing entry. On the other hand the corresponding management operation reports an error. Will we keep this discrepancy also for the new actions and operations?

Reply:
I think keep it.  

IMO the management operation sounds safer as it prevents an accidental overwrite but I am not sure if we want to change the tool in that way at this point - if we did we could probably have an ELY issue opened straight away as it in relation to existing behaviour.

Reply:
Agree.

The note can be updated and this conversation can be resolved.

Reply:
This one is done.